config/v1/types_infrastructure.go

```diff
@@ -505,9 +505,103 @@ type BareMetalPlatformStatus struct {
 	NodeDNSIP string `json:"nodeDNSIP,omitempty"`
 }
 
+type OpenStackAPIBGPPeer struct {
+	// asn is the ASN of the peer.
+	//
+	// +kubebuilder:validation:Required
+	ASN string `json:"asn"`
+
+	// ip is the IP address of the peer. It may be either IPv4 or IPv6.
+	//
+	// +kubebuilder:validation:Required
+	IP string `json:"ip"`
+}
+
+// OpenStackAPIBGPSpeaker species
+type OpenStackAPIBGPSpeaker struct {
+	// failureDomain is the name of a failure domain which this BGP
+	// configuration applies to.
+	//
+	// +kubebuilder:validation:Required
+	FailureDomain string `json:"failureDomain"`
+
+	// asn specifies the ASN to be used by the BGP speaker.
+	//
+	// +kubebuilder:validation:Required
+	ASN string `json:"asn"`
+
+	// peers is a list of all BGP peers of this speaker.
+	//
+	// +kubebuilder:validation:MinItems:=1
+	Peers []OpenStackAPIBGPPeer `json:"peers"`
+}
+
+type OpenStackAPIBGPConfiguration struct {
+	// apiVIPs is a list of IP addresses which will be published via BGP for
+	// use by the API server. It may contain both IPv4 and IPv6 addresses.
+	//
+	// +kubebuilder:validation:MinItems:=1
+	APIVIPs []string `json:"apiVIPs"`
+
+	// speakers is a list of BGP speaker configurations. We require a
+	// speaker configuration for every failure domain hosting a control
+	// plane node.
+	//
+	// +kubebuilder:validation:MinItems:=1
+	// +listType=map
+	// +listMapKey=failureDomain
+	Speakers []OpenStackAPIBGPSpeaker `json:"speakers"`
+}
+
+// OpenStackFailureDomain specifies a failure domain for an OpenStack server.
+// Specifically, it specifies a set of values which will be set on all machines
+// using the failure domain.
+type OpenStackFailureDomain struct {
+	// name is the unique name of this failure domain.
+	//
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+
+	// availabilityZone specifies the OpenStack availability zone for all
+	// servers in this failure domain.
+	//
+	// +kubebuilder:validation:Optional
+	AvailabilityZone string `json:"availabilityZone,omitempty"`
```

Comment:
We should split AvailabilityZone in two for Compute and Storage. Something like:

```Go
	// ComputeZone specifies the OpenStack Compute availability zone for all servers in this failure domain.
	//
	// If not specified, the servers are provisioned in the default Compute
	// availabity zone.
	// +kubebuilder:validation:Optional
	ComputeZone string `json:"computeZone,omitempty"`

	// StorageZone specifies the OpenStack storage availability zone for
	// all volumes in this failure domain.
	//
	// If not specified, the volumes are provisioned in the default storage
	// availabity zone.
	// +kubebuilder:validation:Optional
	StorageZone string `json:"storageZone,omitempty"`

```

Reply:
Agreed.

Reply:
done!